velox/exec/Spill.h

```diff
@@ -0,0 +1,326 @@
+
+  ~SpillFile() override;
+
+  // Returns a file for writing spilled data. The caller constructs
+  // this, then callsoutput() and writes serialized data to the file
+  // and calls finishWrite when the file has reached its final
+  // size. For sorted spilling, the data in one file is expected to b
+  // sorted.
+  WriteFile& output();
+
+  bool isWritable() const {
+    return output_ != nullptr;
+  }
+
+  // Finishes writing and flushes any unwritten data.
+  void finishWrite() {
+    VELOX_CHECK(output_);
+    fileSize_ = output_->size();
+    output_ = nullptr;
+  }
+
+  // Prepares 'this' for reading. Positions the read at the first row of
+  // content. The caller must call output() and finishWrite() before this.
+  void startRead();
+
+  // Returns the file size in bytes. During the writing phase this is
+  // the current size of the file, during reading this is the final
+  // size.
+  uint64_t size() const override {
+    if (output_) {
+      return output_->size();
+    }
+    return fileSize_;
+  }
+
+  // Sets 'result' to refer to the next row of content of 'this'.
+  void read(RowVector& result);
+
+ private:
+  void nextBatch() override;
+
+  const std::string path_;
+  // Byte size of the backing file. Set when finishing writing.
+  uint64_t fileSize_ = 0;
+  std::unique_ptr<WriteFile> output_;
+  std::unique_ptr<SpillInput> input_;
+};
+
+// Sequence of files for one partition of the spilled data. If data is
+// sorted, each file is sorted. The globally sorted order is produced
+// by merging the constituent files.
+class SpillFileList {
+ public:
+  // Constructs a set of spill files. 'type' is a RowType describing the
+  // content. 'numSortingKeys' is the number of leading columns on which the
+  // data is sorted. 'path' is a file path prefix. ' 'targetFileSize' is the
+  // target byte size of a single file in the file set. 'pool' and
+  // 'mappedMemory' are used for buffering and constructing the result data read
+  // from 'this'.
+  //
+  // When writing sorted spill runs, the caller is responsible for buffering and
+  // sorting the data. write is called multiple times, followed by flush().
+  SpillFileList(
+      RowTypePtr type,
+      int32_t numSortingKeys,
+      const std::string& path,
+      uint64_t targetFileSize,
+      memory::MemoryPool& pool,
+      memory::MappedMemory& mappedMemory)
+      : type_(type),
+        numSortingKeys_(numSortingKeys),
+        path_(path),
+        targetFileSize_(targetFileSize),
+        pool_(pool),
+        mappedMemory_(mappedMemory) {}
+
+  // Adds 'rows' for the positions in 'indices' into 'this'. The indices
+  // must produce a view where the rows are sorted if sorting is desired.
+  // Consecutive calls must have sorted data so that the first row of the
+  // next call is not less than the last row of the previous call.
+  void write(
+      const RowVectorPtr& rows,
+      const folly::Range<IndexRange*>& indices);
```

Comment:
Why Range<IndexRange*> and not SelectivityVector? Elsewhere in the codebase SelectivityVector is used to specify a subset of rows in a vector.